Automattic/jetpack
# Connection: add partner codes to connection URLs & deprecate Partner package

## Proposed changes:

This PR changes the way the Partner codes were added when connecting a site to a WordPress.com account, so this can work with standalone plugins.

This was previously only done in the Jetpack plugin, in 2 ways:

- The Partner package is initialized on plugins_loaded in the main Jetpack class.
- In Jetpack's `build_authorize_url` method we have a `jetpack_build_authorize_url` filter that is leveraged by the Partner package.

Standalone plugins, on the other hand, do not use `Jetpack::build_authorize_url()`. This method is obviously only available in the Jetpack plugin, and so standalone plugins do not have that `jetpack_build_authorize_url` filter.

We fix this discrepancy in this PR:

- We now include the Partner package functionality directly in the connection package.
- We initialize the Partner package in the connection package.
- We add the `jetpack_build_authorize_url` filter to the connection package's `get_authorization_url` method, so the partner codes can be added to all URLs.
- We stop initializing the Partner class in the Jetpack plugin, since it is now done in the connection package.

> ~**Note**~
> ~This does not add the Partner package as a new dependency for the Connection package, since the Partner package already declares the connection package as a dependency.~

> **Note**
> This PR is getting larger because we are deprecating the Partner package as part of this PR. It is no longer needed since the features it offered can be offered directly in the connection package.

### Other information:


## Jetpack product discussion

* pbtFPC-6ha-p2

## Does this pull request change what data or activity we track or use?

* No

## Testing instructions:

There may be an easier way to test this, but here is how I went about it.

> **Warning**
> You will want to test the process in both the Jetpack plugin and a standalone plugin like Jetpack Protect

> **Warning**
> It would be useful to test with the Jetpack plugin running this branch, as well as another standalone plugin running a stable version, still using the Partner package.

* On a brand new site,
* Via WP CLI, set up an affiliate code and a subsiadiary code:
    * `wp option update jetpack_partner_subsidiary_id jeremy`
    * `wp option update jetpack_affiliate_code 123456`
* Install a Jetpack plugin (Jetpack or something else).
* Open your browser's network tools
* Filter for `jetpack.wordpress.com` requests
* Start the connection process
* Check the arguments added to the jetpack.wordpress.com request
    * It should include both an `aff` and a `subsidiaryId` argument, using the values defined above.
